### PR TITLE
feat: support ESM in `react-native.config`

### PR DESCRIPTION
**Pull request by szymonrybczak**
Summary:
---------

Closes https://github.com/react-native-community/cli/issues/2167

Add support for ESM inside `react-native.config` file.

Test Plan:
----------

1. Clone the repository and do all the required steps from the [Contributing guide](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md)
2. Create a new project and add `type: "module"` inside package ✅
3. Create a `react-native.config.mjs` with following content

```sh
export default {
  commands: [
    {
      name: 'hello',
      func: () => {
        console.log('hello esm');
      },
    },
  ],
};
```
4. Run `npx react-native hello` and you should see `hello esm` log.

Checklist
----------

- [x] Documentation is up to date to reflect these changes.
- [x] Follows commit message convention described in [CONTRIBUTING.md](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md#commit-message-convention)

### CODE REVIEW

**Review by kraenhansen — APPROVED** [review]

**Comment by satya164** [comment]
Re: test plan, test these scenarios:

- Add `type: 'module'` to `package.json` and use `require` in `react-native.config.js` - should fail unless using `createRequire` helper (`import { createRequire } from 'node:module'; const require = createRequire(import.meta.url);`)
- Add `type: 'module'` to `package.json` and use `require` in `react-native.config.cjs`
- Add `type: 'module'` to `package.json` and use `import/export` in `react-native.config.js`
- Remove `type: 'module` from `package.json`/add `type: 'commonjs'` and use `import/export` in `react-native.config.mjs`

You don’t need both type: 'module' and .mjs. either is enough to signal ES modules.

**Review by TMisiukiewicz — APPROVED** [review]

File: __e2e__/config.test.ts [hunk 1]
```diff
@@ -133,9 +168,110 @@ test('should read user config from react-native.config.js', () => {
...
+test('should read user config from react-native.config.mjs', () => {
+  writeFiles(path.join(DIR, 'TestProject'), {
+    'react-native.config.mjs': USER_CONFIG_ESM,
+  });
+
+  const {stdout} = runCLI(path.join(DIR, 'TestProject'), ['test-command-esm']);
+  expect(stdout).toBe('test-command-esm');
+});
+
+test('should fail if if using require() in ES module in react-native.config.mjs', () => {
```
Comment (thymikee) [inline __e2e__/config.test.ts:187]:
```suggestion
test('should fail if using require() in ES module in react-native.config.mjs', () => {
```

File: __e2e__/config.test.ts [hunk 2]
```diff
@@ -133,9 +168,110 @@ test('should read user config from react-native.config.js', () => {
...
+  const {stderr, stdout} = runCLI(path.join(DIR, 'TestProject'), [
+    'test-command-esm',
+  ]);
+  expect(stderr).toMatch('error Failed to load configuration of your project');
+  expect(stdout).toMatch(
+    'ReferenceError: require is not defined in ES module scope, you can use import instead',
+  );
+});
+
+test('should fail if if using require() in ES module with "type": "module" in package.json', () => {
```
Comment (thymikee) [inline __e2e__/config.test.ts:204]:
```suggestion
test('should fail if using require() in ES module with "type": "module" in package.json', () => {
```

**Comment by thymikee** [comment]
Changing the signature of `loadConfig` would be breaking e.g. for rnx-kit. We could consider creating `loadConfigAsync` or so to avoid it. cc @tido64

**Comment by tido64** [comment]
> Changing the signature of `loadConfig` would be breaking e.g. for rnx-kit. We could consider creating `loadConfigAsync` or so to avoid it. cc @tido64

Thanks, this would indeed break us. I'm also curious why everything needs be async just because of ESM?

**Comment by satya164** [comment]
> everything needs be async just because of ESM

Because it's loading ES module dynamically, which needs to be async.

**Review by thymikee — APPROVED** [review]

**Comment by karlhorky** [comment]
Nice, thanks to all of the contributors and reviewers here, great to have ESM in `react-native.config.js` files!

I'm not super aware of the upstreaming / integration process here in React Native projects - will this work right away out of the box with [`@react-native-community/cli@15.1.0`](https://github.com/react-native-community/cli/releases/tag/v15.1.0)?

Eg:

1. I guess new React Native projects which [start without a framework](https://reactnative.dev/docs/getting-started-without-a-framework) will receive this feature right away now?
2. Does this affect [Expo projects](https://reactnative.dev/docs/environment-setup#start-a-new-react-native-project-with-expo) at all? (I guess not)

**Comment by thymikee** [comment]
Yes, this should work with v15.1. You'll need to update the template (or your project files) to reference the newer version of the CLI. It doesn't affect Expo projects

**Pull request merged**